File: node/keel/src/backend.mjs
/**
 * Backend selection, isolated behind one seam (architecture §5.2, "the swap").
 *
 * Priority:
 *   1. native addon (`keel-core-native`) when loadable — the eventual napi core
 *      (Task 7/14). It is probed by dynamic import and must expose an async
 *      `execute`; it may not exist yet, so failure to load is normal.
 *   2. the in-repo Node backend (AsyncEngine over keel-core-stub semantics).
 *
 * `KEEL_BACKEND` overrides selection:
 *   - `stub`   → force the in-repo engine (never probe native)
 *   - `native` → require the native addon; throw KEEL-E040 if not loadable
 *   - unset    → auto (native if loadable, else engine)
 */

import { join } from "node:path";
import { AsyncEngine, realClock, KeelError } from "./engine.mjs";

const NATIVE_CANDIDATES = [
  "keelrun-core-native", // the addon's actual package name (node/keel-core-native)
  "../../keel-core-native/index.mjs", // in-repo sibling of node/keel (worktree), when built
];

/**
 * Front-end adapter over the napi `KeelCore`. The native surface differs from
 * the front end's contract in one way, bridged here (Task 14 "the swap"):
 *   - the front end drives an ASYNC effect and `await`s `execute`; the native
 *     equivalent is `executeAsync` (tokio↔libuv). We map one to the other.
 * `layer(target, key)` and `resolveTarget(...)` are thin delegations to the
 * native core's own bindings (Task 11/SP-1: the core exposes both directly
 * now, so this wrapper no longer re-derives them from a locally-held policy
 * copy). `persistent` reflects the native journal (the dev-cache
 * `scope=persistent`).
 */
class NativeBackend {
  kind = "native";
  #core;
  constructor(core) {
    this.#core = core;
  }
  configure(policy) {
    this.#core.configure(policy);
  }
  execute(request, effect, idempotencyKey) {
    // `idempotencyKey` (contracts/adapter-pack.md "Idempotency-key injection")
    // matters only while a flow is open — the native `executeAsync` ignores it
    // on the bare-engine branch, so it is always safe to forward.
    return this.#core.executeAsync(request, effect, idempotencyKey); // returns a Promise<Outcome>
  }
  /** Peek the idempotency key recorded for the active flow's next step
   *  (rule 3) — `null` outside a flow or when nothing is recorded, and also
   *  on an addon build too old to expose the native method (optional
   *  chaining, mirroring `flushEvents`). */
  recordedIdempotencyKey(stepKey) {
    return this.#core.recordedIdempotencyKey?.(stepKey) ?? null;
  }
  report() {
    return this.#core.report();
  }
  get persistent() {
    return this.#core.persistent === true;
  }
  // --- Tier 2: durable flows (native-only; absent on the stub) --------------
  // Present only when the loaded addon exposes them (checked once at load
  // time, below) — `_flow.mjs`'s `backendSupportsFlows` probes for these.
  enterFlow(entrypoint, argsHash, { codeHash, explicitKey, leaseMs } = {}) {
    return this.#core.enterFlow(entrypoint, argsHash, codeHash, explicitKey, leaseMs);
  }
  exitFlow(status) {
    this.#core.exitFlow(status);
  }
  journalTime(key, nowMs) {
    return this.#core.journalTime(key, nowMs);
  }
  journalRandom(key, data) {
    return this.#core.journalRandom(key, data);
  }
  layer(target, key) {
    return this.#core.layer(target, key);
  }
  /** Resolve the policy target key for one outbound request — delegates to
   *  the native core's own `resolveTarget` (the LLM host map, Vertex
   *  regional suffix, and `[target]` host/URL-pattern matching,
   *  `docs/targeting.md`), proven identical to the stub backends by
   *  conformance scenarios 36–38. The front ends' `backend.resolveTarget(...)`
   *  reader. */
  resolveTarget(method, host, scheme, port, path) {
    return this.#core.resolveTarget(method, host, scheme, port, path);
  }
  /** Flush the native engine's live NDJSON event feed (`.keel/events/`) —
   * see `KeelCoreNative::flush_events`'s doc comment. A no-op on an addon
   * build too old to expose it. */
  flushEvents() {
    this.#core.flushEvents?.();
  }
}

async function tryLoadNative({ journalPath } = {}) {
  for (const spec of NATIVE_CANDIDATES) {
    let mod;
    try {
      mod = await import(spec);
    } catch {
      continue; // addon not present at this spec — try the next candidate
    }
    const Ctor = mod.KeelCoreNative ?? mod.default;
    if (typeof Ctor !== "function") continue;
    let core;
    try {
      core = journalPath ? new Ctor({ journalPath }) : new Ctor();
    } catch (err) {
      if (!journalPath) continue; // no-journal construction failed — genuinely unusable
      // The addon loaded but the journal could not open (unwritable/invalid path).
      // Degrade to an IN-MEMORY native core rather than downgrading to the stub —
      // resilience still comes from the real engine; only cross-run dev-cache
      // replay is lost. Mirrors the Python front end's graceful fallback.
      process.emitWarning(
        `keel: journal at ${journalPath} could not open (${err?.message ?? err}); ` +
          "native core running in-memory (no cross-run dev cache)",
        { code: "KEEL_JOURNAL_UNAVAILABLE" }
      );
      try {
        core = new Ctor();
      } catch {
        continue;
      }
    }
    if (typeof core.executeAsync === "function" && typeof core.configure === "function") {
      return new NativeBackend(core);
    }
  }
  return null;
}

/**
 * Where the native core attaches its journal (persistent dev cache + Tier 2).
 * `KEEL_JOURNAL` overrides the path; an explicit empty value disables it. This
 * is the *construction-time* default: keel.toml's `journal` key replaces it at
 * configure time unless KEEL_JOURNAL is set, in which case the env wins (see
 * `applyJournalEnvOverride` in bootstrap.mjs).
 */
function resolveJournalPath(cwd, env) {
  const override = env.KEEL_JOURNAL;
  if (override !== undefined) return override === "" ? null : override;
  return join(cwd, ".keel", "journal.db");
}

/**
 * Resolve the runtime backend. Returns an object exposing async
 * `execute(request, effect)`, `configure(policy)`, `layer(target, key)`,
 * `report()`, and `persistent`.
 */
export async function loadBackend({
  preferred = process.env.KEEL_BACKEND,
  clock,
  cwd = process.cwd(),
  env = process.env,
} = {}) {
  // Normalize + validate the selection, matching the Python twin: unset/empty →
  // "auto"; anything other than auto|native|stub is a loud KEEL-E040, never a
  // silent fall-back to auto.
  const choice = (preferred ?? "auto") || "auto";
  if (choice !== "auto" && choice !== "native" && choice !== "stub")
    throw new KeelError("KEEL-E040", `KEEL_BACKEND must be auto|native|stub, got ${JSON.stringify(choice)}`);

  if (choice !== "stub") {
    const native = await tryLoadNative({ journalPath: resolveJournalPath(cwd, env) });
    if (native) return native;
    if (choice === "native")
      throw new KeelError(
        "KEEL-E040",
        "KEEL_BACKEND=native requested but keel-core-native is not loadable"
      );
  }
  return new AsyncEngine(clock ?? realClock());
}

File: python/keel/src/keel/_backend.py
"""Backend selection, isolated behind one seam (architecture §5.1, "the swap").

A backend exposes `configure(policy)`, `execute(request, effect)`, and
`report()` — the four-operation core surface the stub defines (advance_clock
is test-only and unused here). Selection:

  * native (`keel_core`) — the eventual PyO3 module (Task 6/14); probed by
    import and required to expose `configure`/`execute`. It may not exist
    yet, so a failed import is normal.
  * stub (`keel_core_stub.KeelCoreStub`) — the in-repo pure-Python core.

`KEEL_BACKEND` overrides selection:
  * `stub`   → force the stub, never probe native
  * `native` → require `keel_core`; raise KEEL-E040 if not loadable
  * `auto` / unset → native if loadable, else the stub
"""

from __future__ import annotations

import importlib
import os
from pathlib import Path
from typing import Any, Mapping, Protocol, runtime_checkable

from ._errors import KeelError


@runtime_checkable
class Backend(Protocol):
    def configure(self, policy: dict[str, Any]) -> None: ...

    def execute(self, request: dict[str, Any], effect: Any) -> dict[str, Any]: ...

    def report(self) -> dict[str, Any]: ...


class _NativeBackend:
    """Front-end adapter over the native ``keel_core`` core, adding one thing the
    PyO3 surface does not expose: ``layer(target, key)``, resolved from the
    configured policy exactly as the stub's ``_layer`` / Node's
    ``NativeBackend.layer`` do. The adapter packs read ``backend.layer`` to honor
    ``idempotency.header`` and gate cache-body buffering; without this the knob is
    dead under native (a Python/Node parity break). Every other attribute
    (``execute``/``execute_async``/``report``/``persistent``/the flow surface)
    delegates straight through, so the swap is transparent."""

    def __init__(self, core: Any) -> None:
        self._core = core
        self._policy: dict[str, Any] = {}

    def configure(self, policy: dict[str, Any]) -> None:
        self._policy = policy if isinstance(policy, dict) else {}
        self._core.configure(policy)

    def layer(self, target: str, key: str) -> Any:
        t = self._policy.get("target")
        if isinstance(t, dict) and isinstance(t.get(target), dict) and key in t[target]:
            return t[target][key]
        defaults = self._policy.get("defaults")
        if not isinstance(defaults, dict):
            return None
        if target.startswith("llm:"):
            llm = defaults.get("llm")
            if isinstance(llm, dict) and key in llm:
                return llm[key]
        outbound = defaults.get("outbound")
        return outbound.get(key) if isinstance(outbound, dict) else None

    def __getattr__(self, name: str) -> Any:
        # Delegate everything else to the native core (execute, execute_async,
        # report, persistent, enter_flow/exit_flow/journal_*, advance_clock, …).
        return getattr(self._core, name)


def _journal_path(cwd: str | Path | None, env: Mapping[str, str]) -> str | None:
    """Where the native core attaches its journal (persistent dev cache + Tier 2).
    `KEEL_JOURNAL` overrides the path; an explicit empty value disables it."""
    override = env.get("KEEL_JOURNAL")
    if override is not None:
        return override or None  # empty string ⇒ no journal
    base = Path(cwd) if cwd is not None else Path.cwd()
    return str(base / ".keel" / "journal.db")


def _try_load_native(journal_path: str | None) -> Backend | None:
    try:
        mod = importlib.import_module("keel_core")
    except ImportError:
        return None
    ctor = getattr(mod, "KeelCore", None) or getattr(mod, "Core", None)
    if ctor is None:
        return None
    inst = None
    if journal_path:
        try:
            inst = ctor(journal_path=journal_path)  # attaches the persistent journal
        except Exception:
            inst = None  # journal open failed — fall back to an in-memory native core
    if inst is None:
        inst = ctor()
    if callable(getattr(inst, "configure", None)) and callable(getattr(inst, "execute", None)):
        # The native KeelCore exposes configure/execute/execute_async/report/
        # persistent; wrap it only to add `layer(target, key)` (idempotency.header
        # + cache-ttl gate parity), delegating everything else.
        return _NativeBackend(inst)  # type: ignore[return-value]
    return None


def load_backend(
    preferred: str | None = None,
    *,
    cwd: str | Path | None = None,
    env: Mapping[str, str] | None = None,
) -> Backend:
    """Resolve the runtime backend per `KEEL_BACKEND` (or `preferred`). Under a
    native backend, a journal is attached at `<cwd>/.keel/journal.db` (created on
    demand) so the dev cache's `scope=persistent` replays across runs."""
    environ = env if env is not None else os.environ
    choice = (preferred if preferred is not None else environ.get("KEEL_BACKEND", "auto")) or "auto"
    if choice not in ("auto", "native", "stub"):
        # A user env-var mistake, not a Keel bug — E001 (config-level), not E040.
        raise KeelError(
            "KEEL-E001",
            f"KEEL_BACKEND must be one of auto, native, or stub (got {choice!r}); "
            "unset it or correct the value",
        )

    if choice != "stub":
        native = _try_load_native(_journal_path(cwd, environ))
        if native is not None:
            return native
        if choice == "native":
            # A missing build/install is a user-environment problem, not a Keel
            # bug — E001 with a concrete next step, never E040 ("file an issue").
            raise KeelError(
                "KEEL-E001",
                "KEEL_BACKEND=native requires the keel_core native module, which is not "
                "installed; build it (`maturin develop -m crates/keel-py/Cargo.toml`) or "
                "unset KEEL_BACKEND to use the pure-Python stub",
            )

    from keel_core_stub import KeelCoreStub

    return KeelCoreStub()
